// react-native-reanimated-carousel/lib/commonjs/components/Pagination/Basic/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Basic = void 0;

var _reactNative = require("react-native");

var _react = _interopRequireDefault(require("react"));

var _PaginationItem = require("./PaginationItem");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const Basic = props => {
  const {
    activeDotStyle,
    dotStyle,
    progress,
    horizontal = true,
    data,
    size,
    containerStyle,
    renderItem,
    onPress
  } = props;
  if (typeof size === "string" || typeof (dotStyle === null || dotStyle === void 0 ? void 0 : dotStyle.width) === "string" || typeof (dotStyle === null || dotStyle === void 0 ? void 0 : dotStyle.height) === "string") throw new Error("size/width/height must be a number");
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [{
      justifyContent: "space-between",
      alignSelf: "center"
    }, horizontal ? {
      flexDirection: "row"
    } : {
      flexDirection: "column"
    }, containerStyle]
  }, data.map((item, index) => {
    return /*#__PURE__*/_react.default.createElement(_PaginationItem.PaginationItem, {
      key: index,
      index: index,
      size: size,
      count: data.length,
      dotStyle: dotStyle,
      animValue: progress,
      horizontal: !horizontal,
      activeDotStyle: activeDotStyle,
      onPress: () => onPress === null || onPress === void 0 ? void 0 : onPress(index)
    }, renderItem === null || renderItem === void 0 ? void 0 : renderItem(item, index));
  }));
};

exports.Basic = Basic;
