// react-native-page-flipper/lib/module/BookPageBackground.js
import React from 'react';
import { View, StyleSheet } from 'react-native';
import { BookSpine } from './BookPage/BookSpine';

const BookPageBackground = _ref => {
  let {
    left,
    right,
    isFirstPage,
    isLastPage,
    containerSize,
    getPageStyle,
    renderPage,
    renderLastPage,
    shouldRenderLastPage
  } = _ref;
  const leftPageStyle = getPageStyle(false, true);
  const rightPageStyle = getPageStyle(true, true);
  return /*#__PURE__*/React.createElement(View, {
    style: styles.container
  }, /*#__PURE__*/React.createElement(View, {
    style: styles.pageContainer
  }, left && renderPage && /*#__PURE__*/React.createElement(View, {
    style: [leftPageStyle]
  }, renderPage(left)), isFirstPage && /*#__PURE__*/React.createElement(BookSpine, {
    right: false,
    containerSize: containerSize
  })), /*#__PURE__*/React.createElement(View, {
    style: styles.pageContainer
  }, right && renderPage && /*#__PURE__*/React.createElement(View, {
    style: [rightPageStyle]
  }, renderPage(right)), isLastPage && /*#__PURE__*/React.createElement(BookSpine, {
    right: true,
    containerSize: containerSize
  }), shouldRenderLastPage && renderLastPage && /*#__PURE__*/React.createElement(View, {
    style: [rightPageStyle, {
      zIndex: -1
    }]
  }, renderLastPage())));
};

export { BookPageBackground };
const styles = StyleSheet.create({
  pageContainer: {
    flex: 1,
    backfaceVisibility: 'hidden',
    overflow: 'hidden',
    justifyContent: 'center' // backgroundColor: 'white',

  },
  container: {
    position: 'absolute',
    zIndex: -1,
    height: '100%',
    width: '100%',
    flexDirection: 'row'
  }
});
